cassovary-core/src/main/scala/com/twitter/cassovary/graph/MemoryMappedDirectedGraph.scala

```diff
@@ -0,0 +1,153 @@
+package com.twitter.cassovary.graph
+
+import java.io.{BufferedOutputStream, DataOutputStream, File, FileOutputStream}
+
+import com.twitter.cassovary.graph.StoredGraphDir.StoredGraphDir
+import com.twitter.cassovary.util.NodeNumberer
+import com.twitter.cassovary.util.io.{AdjacencyListGraphReader, IntLongSource, MemoryMappedIntLongSource}
+
+/**
+ * A graph which reads edge data from a memory mapped file.  There is no object overhead per node: the memory
+ * used for n nodes and m edges with both in-neighbor and out-neighbor access is exactly 8 + 16*n + 8*m bytes.
+ * Also, Loading is very fast because no parsing of text is required.  Loading time is exactly the time it takes the operating system
+ * to map data from disk into memory.
+ *
+ * Nodes are numbered sequentially from 0 to nodeCount - 1 and must be a range of this form (i.e. nodeCount == maxNodeId + 1).
+ * When transforming a graph where nodeCount <= maxNodeId to this format, new nodes with no neighbors will be implicitly created.
+ *
+ * Currently only supports storing both in-neighbors and out-neighbors of nodes (StoredGraphDir.BothInOut).
+ * The binary format is currently subject to change.
+ */
+
+/* Storage format
+byteCount  data
+4          (reserved, later use for versioning or indicating undirected vs directed)
+4          n (i. e. the number of nodes)
+8*(n+1)    Offsets into out-neighbor data. Index i (a Long) points to the out-neighbor data of node i.
+           Index n is needed to compute the outdegree of node n - 1.
+8*(n+1)    Offsets into in-neighbor data (Longs) (Same interpretation as out-neighbor offsets)
+m          out-neighbor data
```

Comment:
Because of the Long offsets, the out-neighbor data and in-neighbor data can be arbitrarily mixed as long as all neighbors of one direction are together for one node. You might want to point that out.